### PR TITLE
Fix PDB overflow (v2) + Allow PDB to guess atom/residue numbers

### LINKED ISSUE #1945: mdtraj.load() to read various non-standard spec PDB files (ReFix)
I was getting a similar problem before the fix #1849 where if residue numbers were in a Chimera type of set up then I would get an error while trying to load using `traj = mdtraj.load("traj.dcd", top="init.pdb")`. 

However, I got around this by making an additional global switch variable. This worked for me, but additional testing is probably necessary.

The file I changed is here: `<path-to_site-packages>/mdtraj/formats/pdb/pdbstructure.py`

line 54 I added:
```
global _switch
_switch = False
```

Altered these lines as well 133 -144 (simply added switch update and and elif):
```
   try:
        if pdbstructure._next_residue_number > 9999 and num_str in _residue_num_initial_nondecimal_vals:
            # If the next residue number is > 9999 and our current residue number is one of the nondecimal keys,
            # Raise an OverflowError, which will switch to the correct mode to read num_str.
            global _switch
            _switch = True
            raise OverflowError("Need to parse residue number using non-decimal residue modes.")
        elif _switch and pdbstructure._next_residue_number > 9999:
            raise OverflowError("Need to parse residue number using non-decimal residue modes.")
        else:
            #  Within "normal" pdb specifications
            return int(num_str)
```
Hopefully this helps someone else who is having a problem. Let me know if you want me to test this more!

### PR DESCRIPTION
**Pull request by jeremyleung521**
A continuation of the discussion on #1931. Currently mdtraj errors out with ugly ValueErrors if it encounters non-integer residue number/atom number.

Incidentally, if you did not use your pdb as a topology file (i.e. no PdbStructure), you might end up with all 0s as residue numbers (but will load successfully)!

So to deal with the inconsistencies, it will now "guess" the residue numbers and atoms if it overflows in a non-standard way, doesn't exist or if it's rubbish (a str). The residue number and atom numbers will be the previous number + 1.

----

@mattwthompson While looking at the code, there were some inconsistencies in the treatment of residue/atom number errors, so I made it so it'll just guess all the residue/atom numbers (basically lists them sequentially) if it doesn't understand what is passed in the atom/residue number columns. I understand that most residue info could be rubbish, but for atom/residue numbers we can at least do a guess (and maybe raise a warning that we guessed)? If this is not a preferable behavior, I think we could at least throw a more informative error ('Your pdb file is malformed/missing information!') before crashing than just a ValueError with an int-conversion failure.

EDIT:
Now also fixes #1945 with additional test cases.

### CODE REVIEW

**Comment by jeremyleung521** [comment]
Just to be a little more descriptive about what I added to `test_pdb`, I added new atoms after the switch in that are potential false-negatives (e.g., 2711, which is 1001 in hex) and actual hex-only for both atom and residue numbers, which should catch further errors.

**Comment by jeremyleung521** [comment]
@janash If you have the time & could review this PR, that'll be great!

**Comment by andychenbruce** [comment]
I can confirm this PR allows mdtraj to load some trajectories that neither 1.10.0 or 1.10.2 can load,
Example file: [starting_pos_1093_processed.pdb.gz](https://github.com/user-attachments/files/18364944/starting_pos_1093_processed.pdb.gz)

**Comment by mattwthompson** [comment]
I'm in favor of merging but happy to leave this open if others are working on a review (I'm unfortunately not)

**Comment by jeremyleung521** [comment]
> I'm in favor of merging but happy to leave this open if others are working on a review (I'm unfortunately not)

I think @janash said she would look at it sometime soon? The more eyes the better, imo.

**Comment by janash** [comment]
Oop! I had forgotten with the holidays! On my list now!

File: mdtraj/formats/pdb/pdbstructure.py [hunk 1]
```diff
@@ -49,8 +49,96 @@
...
+
+
+_atom_num_initial_nondecimal_functions = {
+    "186a0": (lambda s: int(s, base=16)),  # "hex"
+    "A0000": (lambda s: (int(s[0], base=36) * 10**4 + int(s[1:], base=36))),  # "chimera"
+    "*****": (lambda s: self._next_atom_number),  # "overflow"
+}
+
+_residue_num_initial_nondecimal_functions = {
+    "2710": (lambda s, y=None, z=None: int(s, base=16)),  # 'hex'
```
Comment (janash) [inline mdtraj/formats/pdb/pdbstructure.py:98]: l noticed that all functions in the dictionary have three arguments, but do not use all of them. There are unused parameter `_num_str` in the `_overflow_residue_check` function and unused `y` and `x` in the hex and chimera parsers. From my read of its usage in the code `_read_residue_number`, there isn't a need to have a matching interface for these function, so the extra parameters could be removed. Overflow is only ever accessed explicitly with the "****" key for residue detection

If the functions were called more generically, this would make sense - could be some kind of future-proofing. For now, may be confusing.

**Review by janash — APPROVED** [review]
Looks good to me! I had some minor comments/questions, but overall seems like a nice addition.

I do agree with @jeremyleung521's comment in the description that if atom/residue numbers are guessed it should inform the user with a warning.

File: mdtraj/formats/pdb/pdbstructure.py [hunk 2]
```diff
@@ -49,8 +49,96 @@
...
+            # There is already an atom with this name.
+            return pdbstructure._next_residue_number
+        else:
+            return currentRes.number
+
+
+_atom_num_initial_nondecimal_functions = {
+    "186a0": (lambda s: int(s, base=16)),  # "hex"
+    "A0000": (lambda s: (int(s[0], base=36) * 10**4 + int(s[1:], base=36))),  # "chimera"
+    "*****": (lambda s: self._next_atom_number),  # "overflow"
```
[the remark this answers is not in the document]
Comment (jeremyleung521) [reply mdtraj/formats/pdb/pdbstructure.py:94]: You're correct that self == the corresponding PdbStructure and that the behavior is that when we're in "overflow" mode, we would always want to guess the next atom number. Though I just checked our test and I didn't actually have a test for this, so this might actually break. I'll add something to test this in the next commit.

File: mdtraj/formats/pdb/pdbstructure.py [hunk 3]
```diff
@@ -49,8 +49,96 @@
 
 from mdtraj.core import element
 
-_residue_num_initial_nondecimal_vals = {"A000": "chimera", "2710": "hex", "****": "overflow"}
-_atom_num_initial_nondecimal_vals = {"A0000": "chimera", "186a0": "hex", "*****": "overflow"}
+
+def _overflow_residue_check(num_str, pdbstructure, curr_atom):
+    """
+    Function to check and guess what the current residue is because it's overflowed. Lifted from the
```
[the remark this answers is not in the document]
Comment (jeremyleung521) [reply mdtraj/formats/pdb/pdbstructure.py:55]: There was already logic to read overflowed residue numbers before #1849. I'm just referring to the fact that this function is a repackage/lifted from there.
https://github.com/mdtraj/mdtraj/blob/79f2b2271e22ade2c30c04c4ad91b69bdd47e7ba/mdtraj/formats/pdb/pdbstructure.py#L704-L728

File: mdtraj/formats/pdb/pdbstructure.py [hunk 4]
```diff
@@ -49,8 +49,96 @@
...
+    ):
+        # This is the first residue in the model.
+        return pdbstructure._next_residue_number
+    else:
+        currentRes = pdbstructure._current_model._current_chain._current_residue
+        if currentRes.name_with_spaces != curr_atom.residue_name_with_spaces:
+            # The residue name has changed.
+            return pdbstructure._next_residue_number
+        elif curr_atom.name_with_spaces in currentRes.atoms_by_name:
+            # There is already an atom with this name.
```
[the remark this answers is not in the document]
Comment (jeremyleung521) [reply mdtraj/formats/pdb/pdbstructure.py:85]: This is logic that previously existed before any of the new pdb-reading capabilities. I'm just trying to keep previous behavior intact.

https://github.com/mdtraj/mdtraj/blob/79f2b2271e22ade2c30c04c4ad91b69bdd47e7ba/mdtraj/formats/pdb/pdbstructure.py#L724-L726

Git blame claims that it's done in 2014 to read hex residue numbers 😂 The intent is probably to guess the residue number instead of discarding the atom (because, like you said, atom names need to be unique in a residue).
https://github.com/mdtraj/mdtraj/commit/c7240246ef530601eed21f27e1e51e9070900d0c#diff-8c272c61429f3b979a20351e9fc2dc854202faccac2d2497f5c14616e0185635R724

File: mdtraj/formats/pdb/pdbstructure.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeremyleung521) [reply mdtraj/formats/pdb/pdbstructure.py]: Indeed the reference to "self" is broken and I've since fixed it. Also added a test that would test the overflow "guessing"

File: mdtraj/formats/pdb/pdbstructure.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (janash) [reply mdtraj/formats/pdb/pdbstructure.py:55]: Oh, I got it. The docstring is from a previous commit and refers to something that was done in a previous PR. It's being a bit nit-picky, but I'd be in favor of removing the "lifted from original PDB code down below" as without the context of the commit history the comment doesn't make sense.

**Comment by jeremyleung521** [comment]
I was adding more warnings when guessing + cleaning up the code and caught a small bug, so added another test when we decide to start guessing. Hope this is the last of it.
